src/main/scala/subscriber/PubSubSubscriber.scala

```diff
@@ -0,0 +1,112 @@
+package subscriber
+
+import com.google.api.core.ApiService.{ Listener, State }
+import com.google.api.gax.batching.FlowControlSettings
+import com.google.api.gax.core.InstantiatingExecutorProvider
+import com.google.cloud.pubsub.v1.{ AckReplyConsumer, MessageReceiver, Subscriber => GSubscriber }
+import com.google.pubsub.v1.{ ProjectSubscriptionName, PubsubMessage }
+import org.threeten.bp.Duration
+import zio.blocking._
+import zio.stream.ZStream
+import zio.{ IO, Queue, ZIO, ZManaged }
+
+import java.util.concurrent.TimeUnit
+
+object PubSubSubscriber {
+
+  def subscribe[A](
+      projectId: String,
+      subscription: String,
+      config: PubSubSubscriberConfig,
+      decoder: MessageDecoder[A]
+  ): ZManaged[Blocking, PubSubError, ZStream[Any, Throwable, DecodedMessage[A]]] =
+    for {
+      queue <-
+        Queue
+          .bounded[IO[Throwable, RawMessage]](config.maxOutstandingElementCount)
+          .toManaged(_.shutdown)
+
+      runtime <- ZIO.runtime[Any].toManaged_
+      _ <- createSubscriber(
+        projectId,
+        subscription,
+        config,
+        value =>
+          runtime.unsafeRunAsync_(
+            queue.offer(
+              value
+            )
+          )
+      )
+    } yield ZStream.repeatEffect(takeNextAndDecode(queue, decoder))
```

Comment:
I would probably opt to use `ZStream.async` or `ZStream.asyncM` here instead since it basically does the queuing etc for you.